src/wrapper-esm.js

```diff
@@ -0,0 +1,35 @@
+/*!
+ * jQuery JavaScript Library v@VERSION
+ * https://jquery.com/
+ *
+ * Copyright OpenJS Foundation and other contributors
+ * Released under the MIT license
+ * https://jquery.org/license
+ *
+ * Date: @DATE
+ */
+// For ECMAScript module environments where a proper `window`
+// is present, execute the factory and get jQuery.
+// For environments that do not have a `window` with a `document`
+// (such as Node.js), expose a factory as module.exports.
+// This accentuates the need for the creation of a real `window`.
+// e.g. var jQuery = require("jquery")(window);
+// See ticket trac-14549 for more info.
+var jQueryOrJQueryFactory = typeof window !== "undefined" && window.document ?
+	jQueryFactory( window, true ) :
+	function( w ) {
+		if ( !w.document ) {
+			throw new Error( "jQuery requires a window with a document" );
+		}
+		return jQueryFactory( w );
+	};
```

Comment:
I was thinking whether it'd be better to extract this factory logic to a separate entry point: `jquery/factory`, and have the default one just assume the global `window`. This would simplify these headers.

The current setup doesn't let you freely use the factory in all scenarios anyway - if you load any files depending on jQuery via CommonJS - which is common, through UMD - then that file depends on `"jquery"` resolving to an actual jQuery instance, not a factory. In effect, you need to set the global `window` anyway and the factory remains useless.

What is useful about this wrapper is the `window` encapsulation - if such a global is present, browser globals are all taken from `window` instead of global variables. This is what unlocks libraries like JSDOM. When I mentioned to factory to Domenic some time ago, he didn't even know about the factory way of loading jQuery. 🙃

The problem with an additional entry point is we'd also want to expose the factory in slim version - which would expand from the current two entry points: `jquery` & `jquery/slim` to 4: `jquery`, `jquery/slim`, `jquery/factory`, `jquery/slim-factory`.